File: frontend/app/lib/productAnalytics.ts
import { platformApi } from "../platform/api";

type ProductEventInput = {
  eventName: string;
  campaignId?: string | null;
  properties?: Record<string, string>;
  idempotencyKey?: string | null;
};

export type ProductFeedbackInput = {
  context:
    | "recommendation_usefulness"
    | "explanation_clarity"
    | "forecast_trust"
    | "automation_confidence"
    | "report_quality";
  subjectType: "recommendation" | "explanation" | "forecast" | "automation" | "report";
  subjectId?: string | null;
  campaignId?: string | null;
  rating: number;
  reasonCode:
    | "useful"
    | "clear"
    | "believable"
    | "not_useful_yet"
    | "unclear"
    | "missing_context"
    | "too_technical"
    | "not_believable";
};

export function analyticsDayKey(): string {
  return new Date().toISOString().slice(0, 10).replaceAll("-", "");
}

export async function trackProductEvent({
  eventName,
  campaignId,
  properties = {},
  idempotencyKey,
}: ProductEventInput): Promise<boolean> {
  try {
    await platformApi("/product-analytics/events", {
      method: "POST",
      body: JSON.stringify({
        event_name: eventName,
        campaign_id: campaignId || null,
        properties,
        idempotency_key: idempotencyKey || null,
      }),
    });
    return true;
  } catch {
    // Measurement must never interrupt the owner's work.
    return false;
  }
}

export async function submitProductFeedback(input: ProductFeedbackInput): Promise<void> {
  await platformApi("/product-analytics/feedback", {
    method: "POST",
    body: JSON.stringify({
      context: input.context,
      subject_type: input.subjectType,
      subject_id: input.subjectId || null,
      campaign_id: input.campaignId || null,
      rating: input.rating,
      reason_code: input.reasonCode,
    }),
  });
}

File: backend/app/schemas/product_analytics.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints


SafeIdempotencyKey = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        min_length=1,
        max_length=160,
        pattern=r"^[A-Za-z0-9_.:-]+$",
    ),
]
SafeSubjectId = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        min_length=1,
        max_length=80,
        pattern=r"^[A-Za-z0-9_.:-]+$",
    ),
]


class ProductEventCreateIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    event_name: str = Field(min_length=1, max_length=100)
    campaign_id: str | None = Field(default=None, min_length=36, max_length=36)
    properties: dict[str, Any] = Field(default_factory=dict)
    occurred_at: datetime | None = None
    idempotency_key: SafeIdempotencyKey | None = None


class ProductFeedbackCreateIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    context: Literal[
        "recommendation_usefulness",
        "explanation_clarity",
        "forecast_trust",
        "automation_confidence",
        "report_quality",
    ]
    subject_type: Literal["recommendation", "explanation", "forecast", "automation", "report"]
    subject_id: SafeSubjectId | None = None
    campaign_id: str | None = Field(default=None, min_length=36, max_length=36)
    rating: int = Field(ge=1, le=5)
    reason_code: Literal[
        "useful",
        "clear",
        "believable",
        "not_useful_yet",
        "unclear",
        "missing_context",
        "too_technical",
        "not_believable",
    ]
